ros/src/twist_controller/twist_controller.py

```diff
@@ -1,14 +1,69 @@
+import rospy
+from yaw_controller import YawController
+from lowpass import LowPassFilter
+from pid import PID
 
 GAS_DENSITY = 2.858
 ONE_MPH = 0.44704
 
 
 class Controller(object):
-    def __init__(self, *args, **kwargs):
-        # TODO: Implement
-        pass
-
-    def control(self, *args, **kwargs):
-        # TODO: Change the arg, kwarg list to suit your needs
-        # Return throttle, brake, steer
-        return 1., 0., 0.
+    def __init__(self, vehicle_mass, fuel_capacity, brake_deadband, decel_limit, accel_limit,
+                 wheel_radius, wheel_base, steer_ratio, max_lat_accel, max_steer_angle):
+        min_speed = 0.1
+        self.yaw_controller = YawController(wheel_base, steer_ratio, min_speed, max_lat_accel, max_steer_angle)
+
+        kp = 0.3
+        ki = 0.1
+        kd = 0
+        throttle_min = 0
+        throttle_max = 0.2
+        self.throttle_controller = PID(kp, ki, kd, throttle_min, throttle_max)
+
+        tau = 0.5 # 1/(2pi*tau) = cutoff frequency
+        ts = .02 # sample time
+        self.vel_lpf = LowPassFilter(tau, ts)
+
+        self.vehicle_mass = vehicle_mass
+        self.fuel_capacity = fuel_capacity
+        self.brake_deadband = brake_deadband
+        self.decel_limit = decel_limit
+        self.accel_limit = accel_limit
+        self.wheel_radius = wheel_radius
+
+        self.last_time = rospy.get_time()
+
+    def control(self, dbw_enabled, current_vel, linear_vel, angular_vel):
+        if not dbw_enabled:
+            self.throttle_controller.reset()
+            return 0, 0, 0
+
+        current_vel = self.vel_lpf.filt(current_vel)
+
+        steering = self.yaw_controller.get_steering(linear_vel, angular_vel, current_vel)
+
+        # rospy.logwarn("Target linear vel: %f" % linear_vel)
+        # rospy.logwarn("Target angular vel: %f" % angular_vel)
+        # rospy.logwarn("Current vel: %f" % current_vel)
+        # rospy.logwarn("Filtered current vel: %f" % self.vel_lpf.get())
+        # rospy.logwarn("Steering: %f" % steering)
+
+        vel_error = linear_vel - current_vel
+        self.last_vel = current_vel
+
+        current_time = rospy.get_time()
+        sample_time = current_time - self.last_time
+        self.last_time = current_time
+
+        throttle = self.throttle_controller.step(vel_error, sample_time)
+        brake = 0
+
+        if linear_vel == 0. and current_vel < 0.1:
+            throttle = 0
+            brake = 700
```

Comment:
There is a note to the `DBW Walkthrough` lesson saying that 400 isn't enough.

> Note
In the walkthrough, only 400 Nm of torque is applied to hold the vehicle stationary. This turns out to be slightly less than the amount of force needed, and Carla will roll forward with only 400Nm of torque. To prevent Carla from moving you should apply approximately 700 Nm of torque.